### PR TITLE
Feature/advanced search query formulation support

### PR DESCRIPTION
**Pull request by aristotelisxs**
A first version of the advanced search endpoint, separate to the core_search endpoint.

### Description

We separated the two endpoints (advanced and simple search) here for consistency/maintainability and also because it was more convenient to work with data.frame types for advanced search queries instead of character strings. The parsing of advanced queries checks each "acceptable" filter included in the query and performs conversions to match Elasticsearch query syntax for either including, excluding string terms, performing field specific searches, etc.

### Related Issue

Fix for #4.

### Example

The following will accept a batch request that will perform a search for "data mining" and "machine learning", excluding the terms "social", "science" and "medicine" respectively. Year ranges are also specified as 2013 to 2014 and 2000 to 2016 respectively.

```
query <- data.frame("all_of_the_words" = c("data mining", "machine learning"),
"without_the_words" = c("social science", "medicine"),"year_from" = c("2013","2000"),  
"year_to" = c("2014", "2016"))
```

res <- core_advanced_search(query)
head(res$data)
res$data[[1]]$id

### CODE REVIEW

**Comment by sckott** [comment]
thanks!

the link https://core.ac.uk/docs/#!/all/advanced_search doesn't seem to link to anything for me, where should that link to?

**Comment by sckott** [comment]
- tidy line widths to 80 chars or less 
- include a check that the input is a data.frame, since `apply()` i don't think fails well if e.g. a list is given

**Comment by aristotelisxs** [comment]
The link (https://core.ac.uk/docs/#!/all/advanced_search) is actually not public so I've included a description of the information needed in the data frame for advanced search queries in the details section.

Caught red-handed on the line widths!
Agreed with the is.data.frame check.

**Comment by sckott** [comment]
thanks @aresxs91 -- all done with changes?

**Comment by aristotelisxs** [comment]
Yes, all done, let me know if you spot anything I have missed. Thanks, @sckott

File: R/core_advanced_search.R [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      queries <- create_batch_query_list(parsed_advanced_query, page, limit)
+      args <- NULL
+      
+      core_POST(path = "search", key, args, queries, ...)
+    } else {
+      core_GET(path = file.path("search", parsed_advanced_query), key,
+               list(page = page, pageSize = limit), ...)
+    }
+  } else {
+    print("query must be a data.frame (see details)")
```
Comment (sckott) [inline R/core_advanced_search.R]: I wouldn't use `print()` here. Use `stop()` instead, can can just move that up above the block of code like 

```r
stop("query must be a data.frame (see details)")

parsed_advanced_query <- apply(X = query, MARGIN = 1, FUN = 
                                 parse_advanced_search_query)

if(length(parsed_advanced_query) > 1){
  queries <- create_batch_query_list(parsed_advanced_query, page, limit)
  args <- NULL
  
  core_POST(path = "search", key, args, queries, ...)
} else {
  core_GET(path = file.path("search", parsed_advanced_query), key,
           list(page = page, pageSize = limit), ...)
}
```

File: R/core_advanced_search.R [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#'
+#' @export
+#' @template all
+#' @param query data.frame, required (details for structure)
+#' @param page (character) page number (default: 1), optional
+#' @param limit (character) records to return (default: 10, minimum: 10),
+#' optional
+#' @details `query` should include columns with the following information 
+#' (at least one is required):
+#' 1. `all_of_the_words`: string, with space separated terms that should all exist
```
Comment (sckott) [inline R/core_advanced_search.R]: there's still 5 lines that have widths > 80 chars. including this one

File: R/core_advanced_search.R [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+#' name metadata field
+#' 9. `doi`: string, to be used as an absolute match against the repository 
+#' name metadata field (all other fields will be ignored if included)
+#' 10. `year_from`: string, to filter target document(s) publisher earlier than 
+#' indicated
+#' 11. `year_to`: string, to filter target document(s) publisher later than 
+#' indicated
+#' `core_advanced_search` does the HTTP request and parses, while
+#' `core_advanced_search_` just does the HTTP request, gives back JSON as a 
+#' character string
```
Comment (sckott) [inline R/core_advanced_search.R:37]: add a `@return` tag and describe what is returned

File: R/zzz.R [hunk 4]
```diff
@@ -1,5 +1,192 @@
 core_base <- function() "https://core.ac.uk"
 
+# Advanced search related functions
```
Comment (sckott) [inline R/zzz.R]: - move all these helper functions into a new file with file name indicating what they are, please use `.R` and not `.r`
- be consistent with fxn names, ideally `snake_case`, right now there's combo of camelCase, snake_case and dot.case

File: R/core_advanced_search.R [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#' @examples \dontrun{
+#' query <- data.frame(
+#' "all_of_the_words" = c("data mining", "machine learning"), 
+#' "without_the_words" = c("social science", "medicine"),
+#' "year_from" = c("2013","2000"), 
+#' "year_to" = c("2014", "2016"))
+#' 
+#' res <- core_advanced_search(query)
+#' head(res$data)
+#' res$data[[1]]$id
```
Comment (sckott) [inline R/core_advanced_search.R:47]: you need a closing bracket `}` on a new line after this last examples line

File: R/zzz.R [hunk 6]
```diff
@@ -1,5 +1,192 @@
...
+}
+
+paste3 <- function(..., sep = " ", collapse = NULL, na.rm = T) {
+  if (na.rm == F)
+    paste(..., sep = sep, collapse = collapse)
+  else
+    if (na.rm == T) {
+      paste.na <- function(x, sep) {
+        x <- gsub("^\\s+|\\s+$", "", x)
+        ret <- paste(na.omit(x), collapse = sep)
```
Comment (sckott) [inline R/zzz.R]: namespace all calls to non-base R pkgs like `stats`, `tools`, etc. 

here, namespace like `stats::na.omit` instead of `na.omit`

**Review by sckott — CHANGES_REQUESTED** [review]
looking good, a few more things to change. 

also, can you add some tests in a new test file for the adv. function.

File: R/core_advanced_search.R (hunk 1 above)
Comment (aristotelisxs) [reply R/core_advanced_search.R]: True, missed that while debugging.

File: R/core_advanced_search.R (hunk 3 above)
Comment (aristotelisxs) [reply R/core_advanced_search.R:37]: I haven't seen any other of the functions have a @return tag so I will add it in for them too.

Just to mention that the advanced search return the same data type as the rest, a data frame with:
1) "status" as character, 
2) "totalHits" as integer and
3) "data" as list.

File: R/zzz.R (hunk 4 above)
Comment (aristotelisxs) [reply R/zzz.R]: Agreed, I will move all advanced search related functions.
Will rename fxn_names accordingly

**Comment by aristotelisxs** [comment]
Thanks for you comments again @sckott, as for the advanced search query tests, I guess an approach would be to add a few queries using a random combination of available filters. Would that be enough?

File: R/core_advanced_search.R (hunk 3 above)
Comment (aristotelisxs) [reply R/core_advanced_search.R:37]: Adding to my previous comment, I missed that the references tag would actually cover for what the other endpoints return, whereas the advanced search does not have one. I see why you wanted a return tag here, will add it only for this endpoint.

Comment (sckott) [reply R/core_advanced_search.R:37]: Thanks for adding `@return` (sometimes I forget them too)

that sounds good what you described for return

**Review by sckott — APPROVED** [review]

**Comment by sckott** [comment]
Thanks for adding some tests, we can adjust those as we see fit later. 

Did you do all changes?

**Comment by aristotelisxs** [comment]
I see, great then.

I want to have a closer look at some of the validity checks, not all of them are needed so I'll need some more time on that. Will let you know, thanks!

**Comment by aresxs91** [comment]
I see, great then.

I want to have a closer look at some of the validity checks, not all of them are needed so I'll need some more time on that. Will let you know, thanks!

**Comment by sckott** [comment]
okay

**Comment by aristotelisxs** [comment]
Hi @sckott, after a quick review, there shouldn't be any immediate changes I'll need to do on the filter values validity checks. There may be edge cases not captured but I think for the most part, they should be ready to go. If there's nothing else to add, let's get this into production!

**Comment by sckott** [comment]
okay, looks good

**Pull request merged**